src/util/mpir_pmi.c

```diff
@@ -828,3 +853,69 @@ static void free_pmi_keyvals(INFO_TYPE ** kv, int size, int *counts)
         MPL_free(counts);
     }
 }
+
+static int parse_coord_file(const char *filename)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int i, j, rank;
+    FILE *coords_file;
+
+    coords_file = fopen(filename, "r");
+    MPIR_ERR_CHKANDJUMP1(NULL == coords_file, mpi_errno, MPI_ERR_FILE,
+                         "**filenoexist", "**filenoexist %s", filename);
+
+    /* Skip the first line */
+    fscanf(coords_file, "%*[^\n]\n");
+    MPIR_Process.coords_dims = 0;
+    fscanf(coords_file, "%d:", &rank);
+    while (!feof(coords_file)) {
+        int temp = 0;
+        if (fscanf(coords_file, "%d", &temp) == 1)
+            ++MPIR_Process.coords_dims;
+        else
+            break;
+        if (fgetc(coords_file) == '\n')
+            break;
+    }
+
+    MPIR_Assert(MPIR_Process.coords_dims == 3);
+    rewind(coords_file);
+    /* Skip the first line */
+    fscanf(coords_file, "%*[^\n]\n");
+
+    if (MPIR_Process.coords == NULL) {
+        MPIR_Process.coords =
+            MPL_malloc(MPIR_Process.coords_dims * sizeof(int) * MPIR_Process.size, MPL_MEM_COLL);
+    }
+    memset(MPIR_Process.coords, -1, MPIR_Process.coords_dims * sizeof(int) * MPIR_Process.size);
+
+    for (i = 0; i < MPIR_Process.size; ++i) {
+        int fields_scanned = fscanf(coords_file, "%d:\n", &rank);
+        MPIR_ERR_CHKANDSTMT2(1 != fields_scanned, mpi_errno, MPI_ERR_FILE, goto fn_fail_read,
+                             "**read_file", "**read_file %s %s", filename, strerror(errno));
+        if (rank >= MPIR_Process.size) {
+            if (MPIR_Process.rank == 0)
+                fprintf(stderr, "Warning: rank=%d is outside commsize=%d\n",
+                        rank, MPIR_Process.size);
+            continue;
+        }
+        for (j = 0; j < MPIR_Process.coords_dims; ++j) {
+            /* store the coords in this order: port_id, switch_id, group_id */
+            fields_scanned =
+                fscanf(coords_file, "%d:\n",
```

Comment:
Is this correct?

Reply:
The coords returned by PMIX and the coords in the input file are in this order: switch-group id, switch id, port number, while the MPIR_Process.coords stores them in a reversed order: port number, switch_id, switch-group id.
It may seem a little weird, but keeping the MPIR_Process.coords in the reversed order makes the logic in the topology-aware algorithms clearer and helps with the readability of the algorithms.

Reply:
Fixed the fscanf issue. The original code could work, but it is not the best practice.